src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala

```diff
@@ -169,9 +163,43 @@ class TaskReplaceActor(
 }
 
 object TaskReplaceActor {
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
   val KillComplete = "^TASK_(ERROR|FAILED|FINISHED|LOST|KILLED)$".r
   val FailedToStart = "^TASK_(ERROR|FAILED|LOST|KILLED)$".r
 
   case object RetryKills
+
+  /** Encapsulates the logic how to get a Restart going */
+  private[upgrade] case class RestartStrategy(nrToKillImmediately: Int, maxCapacity: Int)
+
+  private[upgrade] def computeRestartStrategy(app: AppDefinition, runningTasksCount: Int): RestartStrategy = {
+    val minHealthy = (app.instances * app.upgradeStrategy.minimumHealthCapacity).ceil.toInt
+    var maxCapacity = (app.instances * (1 + app.upgradeStrategy.maximumOverCapacity)).toInt
+    var nrToKillImmediately = math.max(0, runningTasksCount - minHealthy)
+
+    if (minHealthy == maxCapacity) {
+      if (app.isResident) {
+        // Kill enough tasks so that we end up with end up with one task below minHealthy.
+        // TODO: We need to do this also while restarting, since the kill could get lost.
+        nrToKillImmediately = runningTasksCount - minHealthy + 1
```

Comment:
`runningTasksCount` could be below `minHealthy`
suggestion: max(runningTasksCount, minHealthy) - minHealthy + 1